src/async.c

```diff
@@ -179,14 +189,17 @@ mpd_async_read(struct mpd_async *async)
 {
 	size_t room;
 	ssize_t nbytes;
+	const size_t min_size = 256;
 
 	assert(async != NULL);
 	assert(async->fd != MPD_INVALID_SOCKET);
 	assert(!mpd_error_is_defined(&async->error));
 
+	if (!mpd_buffer_make_room(&async->input, min_size)) {
```

Comment:
What if you're receiving data from a server which pushes a terabyte to you? DoS bug.

Reply:
This is my understanding, maybe i'm missing something.
I did not change the reading functions to read all received data. It can still fail for large messages. Details below:

`mpd_async_read` is only used when (i) you are going to read the buffer or (ii) writing/flushing output data. In addition, `mpd_buffer_make_room` will only increase the size if there wasn't enough space available.

For (i), you will increase the buffer by 256 bytes (if needed) and read data out. If the message was larger than 256 bytes, an error will occur in `mpd_async_recv_line`. Even if you divide the terabyte message in chunks of 256 bytes, `mpd_async_recv_line` will continually free buffer area, so more allocations will not happen.

For (ii), the flush/write will stop once output data is no longer available. Output data is generated locally. Assuming you wrote 32MB of output data [1] (13 loops on `mpd_sync_send_command_v`) and the input buffer was always full, the new input buffer size is only 7424 bytes.

You can increase the input buffer size by calling `mpd_async_recv_raw`; but this is done deliberately.

[1] MPD will not accept this, but lets say it does for this example.

If you prefer, we could provide in the meson configure the options for the first and limit allocation sizes for the buffer.

Reply:
I think you're wrong. You assume that the message will be delivered in chunks of 256 bytes, but why do you believe that is true?
What if a rogue server sends a single line which is 1 TB long?

Reply:
I see two cases if you send a single 1 TB line
(i) if you are expecting a line, `mpd_async_recv_line` will not be able to read the 1 TB. The maximum size it can increase the input buffer is `min_size`, which is currently 256 bytes. If it does not find `\n` in the input buffer, an error is set.

(ii) You can ask `mpd_async_recv_raw` to receive a 1TB. But this is the user choice. I know Linux does not normally respect this, but this would be the `ENOMEM` case.

Case (ii) is not currently doing this because i forgot to change it. It will only receive up to the current size available.